upstairs/src/block_req.rs

```diff
@@ -78,53 +49,58 @@ impl Drop for BlockRes {
  * When BlockOps are sent to a guest, the calling function receives a waiter
  * that it can block on. This uses a single-use channel to receive the
  * result of a particular operation, and is meant to be paired with a
- * BlockReq.
+ * sender stored in the `BlockOp`.
  */
 #[must_use]
-pub(crate) struct BlockReqWaiter {
-    recv: oneshot::Receiver<BlockReqReply>,
+pub(crate) struct BlockReqWaiter<T, E = CrucibleError> {
+    recv: oneshot::Receiver<Result<T, E>>,
 }
 
-impl BlockReqWaiter {
+impl<T, E> BlockReqWaiter<T, E> {
     /// Create associated `BlockReqWaiter`/`BlockRes` pair
-    pub fn pair() -> (BlockReqWaiter, BlockRes) {
+    pub fn pair() -> (Self, BlockRes<T, E>) {
         let (send, recv) = oneshot::channel();
         (Self { recv }, BlockRes(Some(send)))
     }
 
     /// Consume this BlockReqWaiter and wait on the message
     ///
     /// If the other side of the oneshot drops without a reply, log an error
-    pub async fn wait(self, log: &Logger) -> BlockReqReply {
+    /// and return `None`.
+    pub async fn wait_raw(self, log: &Logger) -> Option<Result<T, E>> {
         match self.recv.await {
-            Ok(reply) => reply,
+            Ok(reply) => Some(reply),
             Err(_) => {
                 warn!(
                     log,
                     "BlockReqWaiter disconnected; \
                      this should only happen at exit"
                 );
```

Comment:
Is it worth doing this logging so deeply in the `BlockReqWaiter`, or leaving it to something higher up to do it inferred from the resulting `RecvDisconnected` error?  It seems a little weird to me to keep a side effect like logging in a result type like this.  *shrug*

Reply:
IIRC this was the compromise position in https://github.com/oxidecomputer/crucible/pull/1085 when we removed "panic if dropped without a reply".  I'd be happy to remove it, though!

Reply:
I guess what I'm asking is: If we choose to continue logging this condition, should it be done this deep in the call stack, considering it has a specific error code (`RecvDisconnected`) dedicated to it, which could be printed later.

Either way, it's not something to get wrapped around the axle over.

Reply:
I think returning the error code is sufficient, and removed the logging in 3a54ada1d1aed63e293555af29097cfae3d0f7ea.